### PR TITLE
executor: run table lookup tasks concurrently if there is no limit.

### PR DESCRIPTION
**Pull request by coocood**
And set index plan concurrency to 10 if OutOfOrder property is set.

### CODE REVIEW

File: executor/executor_xapi.go [hunk 1]
```diff
@@ -332,7 +361,11 @@ func (e *XSelectIndexExec) doIndexRequest() (*xapi.SelectResult, error) {
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	return xapi.Select(txn.GetClient(), selIdxReq, 1)
+	concurrency := 1
```
Comment (shenli) [inline executor/executor_xapi.go]: Why set this to one? The underlying tikv client will ensure the order when using multiple threads.

Comment (coocood) [inline executor/executor_xapi.go]: Index in one region may be enough for the query, set concurrency to 1 to save the work that tikv need to do.

File: executor/executor_xapi.go [hunk 2]
```diff
@@ -170,6 +170,7 @@ type lookupTableTask struct {
 	rows    []*Row
 	cursor  int
 	done    bool
+	respCh  chan error
```
Comment (shenli) [inline executor/executor_xapi.go]: errChan is a better name?

Comment (coocood) [inline executor/executor_xapi.go]: The channel doesn't have to return error, name it errChan is a little confusing.

Comment (shenli) [inline executor/executor_xapi.go]: The name confuses me very much. So I think we should add comment for it.
"respCh" make me believe that the task's response data will be returned from this channel.

File: executor/executor_xapi.go [hunk 3]
```diff
@@ -185,29 +186,25 @@ func (e *XSelectIndexExec) Next() (*Row, error) {
 			return nil, errors.Trace(err)
 		}
 		e.buildTableTasks(handles)
+		if e.indexPlan.NoLimit {
```
Comment (shenli) [inline executor/executor_xapi.go]: If limit > batchSize, we could use concurrent request.

File: executor/executor_xapi.go (hunk 2 above)
Comment (coocood) [inline executor/executor_xapi.go]: How about name it `doneCh`?

Comment (shenli) [inline executor/executor_xapi.go]: Much better.

**Comment by coocood** [comment]
@shenli PTAL

**Comment by shenli** [comment]
LGTM

**Comment by ngaut** [comment]
PTAL @hanfei19910905

**Comment by shenli** [comment]
@zimulala PTAL

File: executor/executor_xapi.go [hunk 4]
```diff
@@ -185,29 +186,32 @@ func (e *XSelectIndexExec) Next() (*Row, error) {
...
-				// Restore the index order.
-				sorter := &rowsSorter{order: e.indexOrder, rows: task.rows}
-				if e.indexPlan.Desc && !e.supportDesc {
-					sort.Sort(sort.Reverse(sorter))
-				} else {
-					sort.Sort(sorter)
+			if task.doneCh != nil {
+				err := <-task.doneCh
+				if err != nil {
+					return nil, err
```
Comment (zimulala) [inline executor/executor_xapi.go]: `err` -> `errors.Trace(err)` ?

File: executor/executor_xapi.go [hunk 5]
```diff
@@ -185,29 +186,32 @@ func (e *XSelectIndexExec) Next() (*Row, error) {
...
+				}
+			} else {
+				err := e.executeTask(task)
```
Comment (zimulala) [inline executor/executor_xapi.go]: Will not repeat execute the same task?

File: executor/executor_xapi.go [hunk 6]
```diff
@@ -332,7 +369,11 @@ func (e *XSelectIndexExec) doIndexRequest() (*xapi.SelectResult, error) {
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	return xapi.Select(txn.GetClient(), selIdxReq, 1)
+	concurrency := 1
+	if e.indexPlan.OutOfOrder {
+		concurrency = 10
```
Comment (zimulala) [inline executor/executor_xapi.go]: Why set this to 10?

Comment (shenli) [inline executor/executor_xapi.go]: It is an empirical value.

File: executor/executor_xapi.go (hunk 5 above)
Comment (coocood) [inline executor/executor_xapi.go]: No, there is a field `done` set after execution.

**Comment by coocood** [comment]
@zimulala PTAL

**Comment by zimulala** [comment]
LGTM

**Pull request merged**